### PR TITLE
Cleanup mongodb tests

### PR DESCRIPTION
**Pull request by anuraaga**
- Use testcontainers to start mongodb to avoid hacky global file lock synchronization
- Use unique collection names to allow sharing mongodb across tests
- Move test code into base class and have subclasses only implement the mongo operations themselves based on that version's API

### CODE REVIEW

File: instrumentation/mongo/mongo-4.0-testing/src/test/groovy/Mongo4ReactiveClientTest.groovy [hunk 1]
```diff
@@ -293,7 +198,9 @@ class Mongo4ReactiveClientTest extends MongoBaseTest {
         "$SemanticAttributes.NET_PEER_IP.key" "127.0.0.1"
```
Comment (mateuszrzeszutek) [inline instrumentation/mongo/mongo-4.0-testing/src/test/groovy/Mongo4ReactiveClientTest.groovy:198]: Can this class use the `mongoSpan` method from the base class?

**Review by mateuszrzeszutek — APPROVED** [review]

File: instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+  def "test collection name for getMore command"() {
+    when:
+    getMore(dbName, collectionName)
+
+    then:
+    assertTraces(2) {
+      trace(0, 1) {
+        mongoSpan(it, 0, "find", collectionName, dbName) {
+          it == '{"find":"' + collectionName + '","filter":{"_id":{"$gte":"?"}},"batchSize":"?"}'
```
Comment (trask) [inline instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy:224]: to match above style?
```suggestion
          assert it == '{"find":"' + collectionName + '","filter":{"_id":{"$gte":"?"}},"batchSize":"?"}'
          true
```

File: instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+        }
+      }
+      trace(1, 1) {
+        mongoSpan(it, 0, "getMore", collectionName, dbName) {
+          it == '{"getMore":"?","collection":"?","batchSize":"?"}'
```
Comment (trask) [inline instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy:229]:
```suggestion
          assert it == '{"getMore":"?","collection":"?","batchSize":"?"}'
          true
```

File: instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+  def "test create collection with already built ClientOptions"() {
+    when:
+    createCollectionWithAlreadyBuiltClientOptions(dbName, collectionName)
+
+    then:
+    assertTraces(1) {
+      trace(0, 1) {
+        mongoSpan(it, 0, "create", collectionName, dbName) {
+          it == "{\"create\":\"$collectionName\",\"capped\":\"?\"}"
```
Comment (trask) [inline instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy:261]:
```suggestion
          assert it == "{\"create\":\"$collectionName\",\"capped\":\"?\"}"
          true
```

File: instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+  // Different client versions have different APIs to do these operations. If adding a test for a new
+  // version, refer to existing ones on how to implement these operations.
+
+  abstract void createCollection(String dbName, String collectionName)
+
+  abstract void createCollectionNoDescription(String dbName, String collectionName)
+
+  // Tests the fix for https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/457
+  // TracingCommandListener might get added multiple times if clientOptions are built using existing clientOptions or when calling  a build method twice.
```
Comment (trask) [inline instrumentation/mongo/mongo-testing/src/main/groovy/AbstractMongoClientTest.groovy:47]:
```suggestion
  // TracingCommandListener might get added multiple times if clientOptions are built using existing clientOptions or when calling a build method twice.
```

File: instrumentation/mongo/mongo-testing/mongo-testing.gradle [hunk 6]
```diff
@@ -2,7 +2,7 @@ apply from: "$rootDir/gradle/java.gradle"
 
 dependencies {
   api project(':testing-common')
-  api group: 'de.flapdoodle.embed', name: 'de.flapdoodle.embed.mongo', version: '1.50.5'
+  api group: 'org.testcontainers', name: 'mongodb', version: versions.testcontainers
```
Comment (trask) [inline instrumentation/mongo/mongo-testing/mongo-testing.gradle:5]: 🎉

**Review by trask — APPROVED** [review]
nice 👍

File: instrumentation/mongo/mongo-4.0-testing/src/test/groovy/Mongo4ReactiveClientTest.groovy (hunk 1 above)
Comment (anuraaga) [reply instrumentation/mongo/mongo-4.0-testing/src/test/groovy/Mongo4ReactiveClientTest.groovy]: Oops, forgot to delete, yeah!

**Pull request merged**